File: pkg/handlerfactory/rustclients/src/fakehandler.rs
use serde::{Deserialize, Serialize};
use std::time::Duration;

use crate::{Client, Result, ClientError};

/// Response from the fake handler
#[derive(Debug, Serialize, Deserialize, Default)]
pub struct FakeResponse {
    #[serde(default)]
    pub message: String,
    #[serde(default)]
    pub status: String,
    #[serde(default)]
    pub code: i32,
}

/// Client for the fake handler
pub struct FakeHandlerClient {
    client: Client,
}

impl FakeHandlerClient {
    /// Create a new fake handler client
    pub fn new(socket_path: &str) -> Self {
        Self {
            client: Client::new(socket_path),
        }
    }
    
    /// Set the connection timeout
    pub fn with_timeout(mut self, timeout: Duration) -> Self {
        self.client = self.client.with_timeout(timeout);
        self
    }
    
    /// Connect to the server
    pub fn connect(&mut self) -> Result<()> {
        self.client.connect()
    }
    
    /// Close the connection
    pub fn close(&mut self) -> Result<()> {
        self.client.close()
    }
    
    /// Return a success message
    pub fn return_success(&mut self, message: Option<&str>) -> Result<String> {
        let mut script = "!!fake.return_success".to_string();
        
        if let Some(msg) = message {
            script.push_str(&format!(" message:'{}'", msg));
        }
        
        self.client.send_command(&script)
    }
    
    /// Return an error message
    pub fn return_error(&mut self, message: Option<&str>) -> Result<String> {
        let mut script = "!!fake.return_error".to_string();
        
        if let Some(msg) = message {
            script.push_str(&format!(" message:'{}'", msg));
        }
        
        // This will return a ClientError::ServerError with the error message
        self.client.send_command(&script)
    }
    
    /// Return a JSON response
    pub fn return_json(&mut self, message: Option<&str>, status: Option<&str>, code: Option<i32>) -> Result<FakeResponse> {
        let mut script = "!!fake.return_json".to_string();
        
        if let Some(msg) = message {
            script.push_str(&format!(" message:'{}'", msg));
        }
        
        if let Some(status_val) = status {
            script.push_str(&format!(" status:'{}'", status_val));
        }
        
        if let Some(code_val) = code {
            script.push_str(&format!(" code:{}", code_val));
        }
        
        let response = self.client.send_command(&script)?;
        
        // Parse the JSON response
        match serde_json::from_str::<FakeResponse>(&response) {
            Ok(result) => Ok(result),
            Err(e) => Err(ClientError::JsonError(e)),
        }
    }
    
    /// Return an invalid JSON response
    pub fn return_invalid_json(&mut self) -> Result<FakeResponse> {
        let script = "!!fake.return_invalid_json";
        let response = self.client.send_command(&script)?;
        
        // This should fail with a JSON parsing error
        match serde_json::from_str::<FakeResponse>(&response) {
            Ok(result) => Ok(result),
            Err(e) => Err(ClientError::JsonError(e)),
        }
    }
    
    /// Return an empty response
    pub fn return_empty(&mut self) -> Result<String> {
        let script = "!!fake.return_empty";
        self.client.send_command(&script)
    }
    
    /// Return a large response
    pub fn return_large(&mut self, size: Option<i32>) -> Result<String> {
        let mut script = "!!fake.return_large".to_string();
        
        if let Some(size_val) = size {
            script.push_str(&format!(" size:{}", size_val));
        }
        
        self.client.send_command(&script)
    }
    
    /// Return a malformed error message
    pub fn return_malformed_error(&mut self) -> Result<String> {
        let script = "!!fake.return_malformed_error";
        self.client.send_command(&script)
    }
    
    /// Get help information
    pub fn help(&mut self) -> Result<String> {
        let script = "!!fake.help";
        self.client.send_command(&script)
    }
}

File: pkg/api/handlerfactory/fakehandler/fake_handler.go
package fakehandler

import (
	"encoding/json"
	"fmt"

	"github.com/freeflowuniverse/herolauncher/pkg/handlerfactory/core"
)

// FakeResponse represents a fake response for testing
type FakeResponse struct {
	Message string `json:"message"`
	Status  string `json:"status"`
	Code    int    `json:"code"`
}

// FakeHandler is a handler for testing different response types
type FakeHandler struct {
	core.BaseHandler
}

// NewFakeHandler creates a new fake handler
func NewFakeHandler() *FakeHandler {
	return &FakeHandler{
		BaseHandler: core.BaseHandler{
			ActorName: "fake",
		},
	}
}

// ReturnSuccess returns a success message
func (h *FakeHandler) ReturnSuccess(script string) string {
	params, err := h.ParseParams(script)
	if err != nil {
		return fmt.Sprintf("Error parsing parameters: %v", err)
	}

	message := params.Get("message")
	if message == "" {
		message = "Success message"
	}
	return message
}

// ReturnError returns an error message
func (h *FakeHandler) ReturnError(script string) string {
	params, err := h.ParseParams(script)
	if err != nil {
		return fmt.Sprintf("Error parsing parameters: %v", err)
	}

	message := params.Get("message")
	if message == "" {
		message = "This is a test error"
	}
	return fmt.Sprintf("Error: %s", message)
}

// ReturnJSON returns a JSON response
func (h *FakeHandler) ReturnJSON(script string) string {
	params, err := h.ParseParams(script)
	if err != nil {
		return fmt.Sprintf("Error parsing parameters: %v", err)
	}

	message := params.Get("message")
	if message == "" {
		message = "JSON response"
	}

	status := params.Get("status")
	if status == "" {
		status = "ok"
	}
	code := params.GetIntDefault("code", 200)

	response := FakeResponse{
		Message: message,
		Status:  status,
		Code:    code,
	}

	jsonBytes, err := json.Marshal(response)
	if err != nil {
		return fmt.Sprintf("Error marshaling JSON: %v", err)
	}

	return string(jsonBytes)
}

// ReturnInvalidJSON returns an invalid JSON response
func (h *FakeHandler) ReturnInvalidJSON(script string) string {
	return "{\"message\": \"This is invalid JSON, missing closing brace"
}

// ReturnEmpty returns an empty response
func (h *FakeHandler) ReturnEmpty(script string) string {
	return ""
}

// ReturnLarge returns a large response
func (h *FakeHandler) ReturnLarge(script string) string {
	params, err := h.ParseParams(script)
	if err != nil {
		return fmt.Sprintf("Error parsing parameters: %v", err)
	}

	size := params.GetIntDefault("size", 1000)
	if size > 10000 {
		size = 10000 // Limit to 10000 to prevent abuse
	}

	response := ""
	for i := 0; i < size; i++ {
		response += fmt.Sprintf("Line %d: This is a test line\n", i+1)
	}

	return response
}

// ReturnMalformedError returns a malformed error message
func (h *FakeHandler) ReturnMalformedError(script string) string {
	return "error: this is not in the standard format"
}

// Help returns help information
func (h *FakeHandler) Help(script string) string {
	return `Fake Handler Commands:
- fake.return_success [message:'custom message'] - Returns a success message
- fake.return_error [message:'custom error'] - Returns an error message
- fake.return_json [message:'custom message'] [status:'custom status'] [code:123] - Returns a JSON response
- fake.return_invalid_json - Returns an invalid JSON response
- fake.return_empty - Returns an empty response
- fake.return_large [size:1000] - Returns a large response with the specified number of lines
- fake.return_malformed_error - Returns a malformed error message
- fake.help - Shows this help message`
}
